notebooks/02-weather_delays.ipynb

```diff
@@ -0,0 +1,669 @@
+   "metadata": {},
+   "source": [
+    "##### **Average percentage in a year**"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_average_df(years, df):\n",
+    "    rows = df.filter(F.col('Year').isin(*years)) \\\n",
+    "             .groupBy('Year') \\\n",
+    "             .avg('WeeklyWeatherDelays') \\\n",
+    "             .withColumnRenamed('avg(WeeklyWeatherDelays)', 'AverageWeatherDelaysPercentage') \\\n",
+    "             .select('Year', 'AverageWeatherDelaysPercentage') \\\n",
+    "             .collect()\n",
+    "    \n",
+    "    nb_years = len(years)\n",
+    "    data = np.zeros(nb_years)\n",
+    "    for row in rows:\n",
+    "        year = row[0] - years[0]\n",
+    "        avg_pen = row[1]\n",
+    "        data[year] = avg_pen \n",
+    "    res = pd.DataFrame({'Weather delays': data}, index=years)\n",
+    "    return res\n",
+    "\n",
+    "def plot_average_weather_delays(years, df, ax):\n",
+    "    df = get_average_df( years, df)\n",
+    "    title = 'Average weather delays percentage'\n",
+    "    if df.empty:\n",
+    "        print('No data')\n",
+    "    else:\n",
+    "        df.plot.bar( title=title, rot=90, ax=ax)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def ui_callback( years, df):\n",
+    "    plt.figure(figsize=(9,12))\n",
+    "    plt.clf()\n",
+    "    ax = plt.subplot(211)\n",
+    "    plot_weather_time_series(range(years[0], years[1] + 1), df, ax)\n",
+    "    \n",
+    "    ax = plt.subplot(212)\n",
+    "    plot_average_weather_delays(range(years[0], years[1] + 1), df, ax)\n",
+    "    \n",
+    "    plt.subplots_adjust(top=0.92, bottom=0.08, left=0.10, right=0.95, hspace=0.25,\n",
+    "                    wspace=0.35)\n",
+    "    plt.show()\n",
+    "\n",
+    "\n",
+    "# Years selection range\n",
+    "years = range(1994, 2009)\n",
+    "years = [(str(y), y) for y in years]\n",
+    "years_w = widgets.SelectionRangeSlider(options=years,\n",
+    "                                       index=(0, 2),\n",
+    "                                       description='Years',\n",
+    "                                       continuous_update=False)\n",
+    "#We could improve the visualization allowing to select the delay type percentage. \n",
+    "#Now can be visualized only the weather delays percentage\n",
+    "\n",
+    "ui = widgets.HBox([years_w])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 18,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "application/vnd.jupyter.widget-view+json": {
+       "model_id": "b80310b1a6974810a4681a01f3a9774a",
+       "version_major": 2,
+       "version_minor": 0
+      },
+      "text/plain": [
+       "HBox(children=(SelectionRangeSlider(continuous_update=False, description='Years', index=(0, 2), options=(('199…"
```

Comment:
Given that the weather data is available only from 2003 onwards, I'd suggest to set at least the default starting position for the slider from 2003.

Reply:
Yes, good point! Do you also suggest to visualize only data after 2003 or to keep the possibility to visualize also data before 2003?

Reply:
Maybe leave the possibility to do it, just in case the professor wants to see if everything is working anyway. :+1:

Reply:
Ok, got it :)